swap/src/asb/config.rs

```diff
@@ -11,14 +13,69 @@ use serde::{Deserialize, Serialize};
 use std::ffi::OsStr;
 use std::fs;
 use std::path::{Path, PathBuf};
+use std::str::FromStr;
 use tracing::info;
 use url::Url;
 
-const DEFAULT_LISTEN_ADDRESS_TCP: &str = "/ip4/0.0.0.0/tcp/9939";
-const DEFAULT_LISTEN_ADDRESS_WS: &str = "/ip4/0.0.0.0/tcp/9940/ws";
-const DEFAULT_ELECTRUM_RPC_URL: &str = "ssl://electrum.blockstream.info:60002";
-const DEFAULT_MONERO_WALLET_RPC_TESTNET_URL: &str = "http://127.0.0.1:38083/json_rpc";
-const DEFAULT_BITCOIN_CONFIRMATION_TARGET: usize = 3;
+pub trait GetDefaults {
+    fn getConfigFileDefaults() -> Result<Defaults>;
+}
+
+pub struct Defaults {
+    pub config_path: PathBuf,
+    data_dir: PathBuf,
+    listen_address_tcp: Multiaddr,
+    listen_address_ws: Multiaddr,
+    electrum_rpc_url: Url,
+    monero_wallet_rpc_url: Url,
+    bitcoin_confirmation_target: usize,
+}
+
+impl GetDefaults for Testnet {
+    fn getConfigFileDefaults() -> Result<Defaults> {
+        let defaults = Defaults {
+            config_path: system_config_dir()
+                .map(|dir| Path::join(&dir, "asb"))
+                .map(|dir| Path::join(&dir, "testnet"))
+                .map(|dir| Path::join(&dir, "config.toml"))
+                .context("Could not generate default config file path")?,
+            data_dir: system_data_dir()
+                .map(|proj_dir| Path::join(&proj_dir, "asb"))
+                .map(|proj_dir| Path::join(&proj_dir, "testnet"))
+                .context("Could not generate default data dir")?,
+            listen_address_tcp: Multiaddr::from_str("/ip4/0.0.0.0/tcp/9939")?,
+            listen_address_ws: Multiaddr::from_str("/ip4/0.0.0.0/tcp/9940/ws")?,
+            electrum_rpc_url: Url::parse("ssl://electrum.blockstream.info:60002")?,
+            monero_wallet_rpc_url: Url::parse("http://127.0.0.1:38083/json_rpc")?,
+            bitcoin_confirmation_target: 1,
+        };
+
+        Ok(defaults)
+    }
+}
+
+impl GetDefaults for Mainnet {
+    fn getConfigFileDefaults() -> Result<Defaults> {
+        let defaults = Defaults {
+            config_path: system_config_dir()
+                .map(|dir| Path::join(&dir, "asb"))
+                .map(|dir| Path::join(&dir, "mainnet"))
+                .map(|dir| Path::join(&dir, "config.toml"))
+                .context("Could not generate default config file path")?,
```

Comment:
If you move the `.context` up onto `system_config_dir`, you can just use `.join` for all the remaining calls and don't have to do this noisy `.map`.